File: scripts/generate-dist.py
#!/usr/bin/env python3
"""Generate per-agent files from portable rule markdown.

Cursor gets always-on .mdc files. Codex concatenation is a later PR.
"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

FRONTMATTER_RE = re.compile(r"\A---\r?\n(.*?)\r?\n---(?:\r?\n)?(.*)\Z", re.DOTALL)
ALWAYS_APPLY_RE = re.compile(r"^alwaysApply\s*:")
README_NAMES = {"readme.md"}


def parse_rule(text: str) -> tuple[str | None, str]:
    match = FRONTMATTER_RE.match(text)
    if not match:
        return None, text
    return match.group(1), match.group(2)


def cursor_frontmatter(raw: str | None) -> str:
    lines: list[str] = []
    if raw:
        for line in raw.splitlines():
            if ALWAYS_APPLY_RE.match(line):
                continue
            lines.append(line)
    lines.append("alwaysApply: true")
    return "---\n" + "\n".join(lines) + "\n---\n"


def render_cursor_rule(text: str) -> str:
    frontmatter, body = parse_rule(text)
    body = body.lstrip("\n")
    if body and not body.endswith("\n"):
        body += "\n"
    return cursor_frontmatter(frontmatter) + ("\n" + body if body else "\n")


def generate_cursor_rules(source_dir: Path, dest_dir: Path) -> list[Path]:
    if not source_dir.is_dir():
        raise FileNotFoundError(f"rules directory not found: {source_dir}")

    dest_dir.mkdir(parents=True, exist_ok=True)
    for old in dest_dir.glob("*.mdc"):
        old.unlink()

    written: list[Path] = []
    for src in sorted(source_dir.glob("*.md")):
        if src.name.lower() in README_NAMES:
            continue
        dest = dest_dir / f"{src.stem}.mdc"
        dest.write_text(render_cursor_rule(src.read_text(encoding="utf-8")), encoding="utf-8")
        written.append(dest)
    return written


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--root",
        type=Path,
        default=Path(__file__).resolve().parent.parent,
        help="Repository root (default: parent of scripts/)",
    )
    parser.add_argument(
        "--cursor-out",
        type=Path,
        default=None,
        help="Cursor rules output directory (default: <root>/dist/cursor/rules)",
    )
    args = parser.parse_args(argv)

    root = args.root.resolve()
    cursor_out = (args.cursor_out or root / "dist" / "cursor" / "rules").resolve()
    written = generate_cursor_rules(root / "rules", cursor_out)
    print(f"Wrote {len(written)} Cursor rule(s) to {cursor_out}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/generate-dist.mjs
#!/usr/bin/env node

import { mkdir, readdir, readFile, unlink, writeFile } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const FRONTMATTER_RE = /^---\r?\n(.*?)\r?\n---(?:\r?\n)?(.*)$/s;
const ALWAYS_APPLY_RE = /^alwaysApply\s*:/;
const README_NAMES = new Set(['readme.md']);

function parseArgs(argv) {
  const args = { root: undefined, cursorOut: undefined, codexOut: undefined };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === '--root') {
      args.root = argv[i + 1];
      i += 1;
    } else if (arg === '--cursor-out') {
      args.cursorOut = argv[i + 1];
      i += 1;
    } else if (arg === '--codex-out') {
      args.codexOut = argv[i + 1];
      i += 1;
    } else {
      throw new Error(`Unknown argument: ${arg}`);
    }
  }
  return args;
}

function parseRule(text) {
  const match = text.match(FRONTMATTER_RE);
  if (!match) {
    return { frontmatter: null, body: text };
  }
  return { frontmatter: match[1], body: match[2] };
}

function cursorFrontmatter(raw) {
  const lines = [];
  if (raw) {
    for (const line of raw.split(/\r?\n/)) {
      if (!ALWAYS_APPLY_RE.test(line)) {
        lines.push(line);
      }
    }
  }
  lines.push('alwaysApply: true');
  return `---\n${lines.join('\n')}\n---\n`;
}

function renderCursorRule(text) {
  const parsed = parseRule(text);
  let body = parsed.body.replace(/^\n+/, '');
  if (body && !body.endsWith('\n')) {
    body += '\n';
  }
  return cursorFrontmatter(parsed.frontmatter) + (body ? `\n${body}` : '\n');
}

function renderCodexBody(text) {
  return parseRule(text).body.replace(/^\n+/, '').replace(/\n+$/, '');
}

async function listRuleFiles(sourceDir) {
  let entries;
  try {
    entries = await readdir(sourceDir, { withFileTypes: true });
  } catch (error) {
    if (error && error.code === 'ENOENT') {
      throw new Error(`rules directory not found: ${sourceDir}`);
    }
    throw error;
  }

  return entries
    .filter((item) => item.isFile() && item.name.endsWith('.md'))
    .filter((item) => !README_NAMES.has(item.name.toLowerCase()))
    .sort((a, b) => a.name.localeCompare(b.name));
}

async function generateCursorRules(sourceDir, destDir) {
  const entries = await listRuleFiles(sourceDir);
  await mkdir(destDir, { recursive: true });
  const existing = await readdir(destDir);
  await Promise.all(
    existing
      .filter((name) => name.endsWith('.mdc'))
      .map((name) => unlink(join(destDir, name))),
  );

  const written = [];
  for (const entry of entries) {
    const dest = join(destDir, `${entry.name.slice(0, -3)}.mdc`);
    const text = await readFile(join(sourceDir, entry.name), 'utf8');
    await writeFile(dest, renderCursorRule(text));
    written.push(dest);
  }
  return written;
}

async function generateCodexAgents(sourceDir, destFile) {
  const entries = await listRuleFiles(sourceDir);
  const bodies = [];
  for (const entry of entries) {
    const text = await readFile(join(sourceDir, entry.name), 'utf8');
    const body = renderCodexBody(text);
    if (body) {
      bodies.push(body);
    }
  }

  await mkdir(dirname(destFile), { recursive: true });
  await writeFile(
    destFile,
    bodies.length > 0 ? `${bodies.join('\n\n')}\n` : '',
  );
  return destFile;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const here = dirname(fileURLToPath(import.meta.url));
  const root = resolve(args.root ?? join(here, '..'));
  const rulesDir = join(root, 'rules');
  const cursorOut = resolve(
    args.cursorOut ?? join(root, 'dist', 'cursor', 'rules'),
  );
  const codexOut = resolve(args.codexOut ?? join(root, 'dist', 'codex'));
  const written = await generateCursorRules(rulesDir, cursorOut);
  await mkdir(codexOut, { recursive: true });
  await generateCodexAgents(rulesDir, join(codexOut, 'AGENTS.md'));
  console.log(`Wrote ${written.length} Cursor rule(s) to ${cursorOut}`);
  console.log(`Wrote Codex dist to ${codexOut}`);
}

const invokedDirectly =
  process.argv[1] &&
  fileURLToPath(import.meta.url) === resolve(process.argv[1]);

if (invokedDirectly) {
  main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
  });
}
